File: NOH_HYE_MIN/Programmers/VisitedLength.java
// U, R 경우에 pos 위치 update 되는 타이밍때문에 수행시간이 많이 달라짐 ㅇㅅㅇ....

class Solution {
    public int solution(String dirs) {
        int answer = 0;
        boolean[][] boardX = new boolean[11][11];
        boolean[][] boardY = new boolean[11][11];
        int[] pos = {5, 5};
        String[] dirsArray = dirs.split("");
        for(int i=0; i<dirsArray.length; i++){
            switch(dirsArray[i]){
                case "U":
                    if(pos[1]<10){
                        if(!boardX[pos[0]][pos[1]]) answer++;
                        boardX[pos[0]][pos[1]] = true;
                        pos[1]=pos[1]+1;
                    }
                    break;
                case "D":
                    if(pos[1]>0){
                        pos[1]=pos[1]-1;
                        if(!boardX[pos[0]][pos[1]]) answer++;
                        boardX[pos[0]][pos[1]] = true;
                    }
                    break;
                case "R":
                    if(pos[0]<10){
                        if(!boardY[pos[0]][pos[1]]) answer++;
                        boardY[pos[0]][pos[1]] = true;
                        pos[0]=pos[0]+1;
                    }
                    break;
                case "L":
                    if(pos[0]>0){
                        pos[0]=pos[0]-1;
                        if(!boardY[pos[0]][pos[1]]) answer++;
                        boardY[pos[0]][pos[1]] = true;
                    }
                    break;
            }
        }
        return answer;
    }
}

File: JANG_SO_HYEON/Programmers/VisitedLength.py
dx = [0, 0, -1, 1] # 위, 아래, 오른쪽, 왼쪽
dy = [1, -1, 0, 0]

d = {"U": 0, "D": 1, "L": 2, "R": 3}

def solution(dirs):
    answer = 0
    x, y = 0, 0
    location = set()
    
    for dir in dirs:
        i = d[dir]
        
        nx, ny = x + dx[i], y + dy[i]
        
        # 좌표평면
        if nx < -5 or nx > 5 or ny < -5 or ny > 5:
            continue
        
        if (x, y, nx, ny) not in location:
            # 이미 지나간 길(양방향)
            location.add((x, y, nx, ny))
            location.add((nx, ny, x, y))
            answer += 1
        
        # 이전의 좌표에서 현재 움직인 좌표로 바꿔주기
        x, y = nx, ny
    
    return answer
